source/extensions/upstreams/http/udp/upstream_request.cc

```diff
@@ -0,0 +1,141 @@
+#include "source/extensions/upstreams/http/udp/upstream_request.h"
+
+#include <cstdint>
+#include <memory>
+
+#include "envoy/upstream/upstream.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/codes.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/message_impl.h"
+#include "source/common/network/transport_socket_options_impl.h"
+#include "source/common/router/router.h"
+#include "source/extensions/common/proxy_protocol/proxy_protocol_header.h"
+
+#include "quiche/common/masque/connect_udp_datagram_payload.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Upstreams {
+namespace Http {
+namespace Udp {
+
+void UdpConnPool::newStream(Router::GenericConnectionPoolCallbacks* callbacks) {
+  Router::UpstreamToDownstream& upstream_to_downstream = callbacks->upstreamToDownstream();
+  Network::SocketPtr socket = createSocket(host_);
+  const Network::ConnectionInfoProvider& connection_info_provider =
+      socket->connectionInfoProvider();
+  ASSERT(upstream_to_downstream.connection().has_value());
+  Event::Dispatcher& dispatcher = upstream_to_downstream.connection()->dispatcher();
+  auto upstream =
+      std::make_unique<UdpUpstream>(&upstream_to_downstream, std::move(socket), host_, dispatcher);
+  StreamInfo::StreamInfoImpl stream_info(dispatcher.timeSource(), nullptr);
+  callbacks->onPoolReady(std::move(upstream), host_, connection_info_provider, stream_info, {});
+}
+
+UdpUpstream::UdpUpstream(Router::UpstreamToDownstream* upstream_to_downstream,
+                         Network::SocketPtr socket, Upstream::HostConstSharedPtr host,
+                         Event::Dispatcher& dispatcher)
+    : upstream_to_downstream_(upstream_to_downstream), socket_(std::move(socket)), host_(host),
+      dispatcher_(dispatcher) {
+  socket_->ioHandle().initializeFileEvent(
+      dispatcher_, [this](uint32_t) { onSocketReadReady(); }, Event::PlatformDefaultTriggerType,
+      Event::FileReadyType::Read);
+}
+
+void UdpUpstream::encodeData(Buffer::Instance& data, bool /*end_stream*/) {
```

Comment:
What do `data` and `end_stream` represent here? I was assuming that `data` was a chunk of the capsule stream and `end_stream` means the capsule stream is over, but if that's the case we should only call `capsule_parser_.ErrorIfThereIsRemainingBufferedData()` once `end_stream` is set